running-code/execute-package.md

```diff
@@ -0,0 +1,163 @@
+cannot be directly executed. But the error message contains the hint that you need to make this run properly.
+
+[Earlier you learned](execute-script-name-eq-main) that `if __name__ == "__main__":` can protect parts of your
+script from executing when it is imported, making that conditional change the file's behavior when used as a script.
+There is a very similar concept that can be applied to Python packages.
+
+Any package that contains a [`__main__.py` module](https://docs.python.org/3/library/__main__.html#module-__main__)
+can be executed directly using the `python` command, without reference to any specific module files.
+
+:::{note}
+The `__main__.py` file typically doesn't have an `if __name__ == "__main__":` conditional in it, as its execution
+is already separated out from the rest of the package.
+:::
+
+### Further exploration
+
+- Try to separate `my_program` into a package containing two files, including one `__main__.py`
+- Try to get `python -m my_program` to work as before
+- Does importing `my_program` still work [as before the separation](execute-script-name-eq-main)?
+
+```python
+import my_program
+
+def guess_my_number():
+    my_program.shiny_hello()
+    print("Was your number 42?")
+
+guess_my_number()
+# ✨ Hello from Python ✨
+# Was your number 42?
+```
+
+:::{attention}
+Don't forget to (re)install your package after creating this file!
+:::
+
+## 2. Named Commands
+
+The final way to make Python code executable directly from the command line is to include a special entrypoint
+into the package metadata. Entrypoints are a general purpose plug-in system for Python packages, but the
+[`console_scripts`](https://packaging.python.org/en/latest/specifications/entry-points/#use-for-scripts)
+entry is specifically targeted at creating executable commands on systems that install the package.
+
+These commands are configured in your project's [`pyproject.toml`](https://www.pyopensci.org/python-package-guide/tutorials/pyproject-toml.html#what-is-a-pyproject-toml-file).
+
+```toml
+[project.scripts]
+shiny = "my_program.greetings:shiny_hello"
+```
+
+In the above example `shiny` is the name of the command that will be made available in the shell after installation,
+`my_program` is the name of your top-level package import, `greetings` is the name of the sub-package (optional, or may be
+repeated as necessary to access the correct sub-package), and `shiny_hello` is the function that will be called.
+
+If this package was installed, the command would be made avalibel in your shell
+
+```bash
+shiny
+# ✨ Hello from Python ✨
+```
+
+The target of each `scripts` definition should always be one function within your package, which will be directly executed (without arguments)
+when the command is invoked in the shell. The target function can live anywhere; it does not have to be in a `__main__.py` or under a `if __name__ == "__main__":`.
+
```

Comment:
i think we probably need a little admonition box about `project.scripts` vs `project.gui-scripts` on windows, at least linking back to the docs.

edit: nvm - since this is a lesson rather than a reference guide, we don't need to cover everything. maybe put that in the "further exploration"

Reply:
But then I'd feel compelled to explain `.pyw` and `pythonw.exe` and something about window management and, and...